Refactored the control panel and workspace reload

JensHouses opened the pull request:
should fix the issues:
  **_QoL improvement: submit/next/prev buttons on navbar #105_** 
and
 **_Better reconnection for nonvnc on challenge restart #478_** 

- I Used BroadcastChannel to trigger an event through browser tabs and/or windows if a challenge (re)started.
- VSCode reloads the page.
- noVNC changes the iframe src.
- I added a description to the control panel and moved the notifications to the bottom of the control panel. I Limited the max-size to 50vh and made it scrollable. Therefor I changed the scrollbar to be more visible.
- I added the description to the test of active_module.

ConnorNelson commented on dojo_theme/static/css/custom.css:
```diff
@@ -17,6 +17,15 @@ body, table, .form-control:focus, .close {
     justify-content: space-evenly;
 }
+html {
+    ::-webkit-scrollbar {
+        width: 5px;
+    }
+    ::-webkit-scrollbar-thumb {
+        background: #b5e853;
+        border-radius: 5px;
+    }
```
I see your comment now. How important is this?

JensHouses replied:
not important at all :-) i'll remove it

JensHouses replied:
https://github.com/pwncollege/dojo/pull/489/commits/3ca1300ebc19fbcf72c6dc1abed4b6e57f32b599 removes it :-)

JensHouses commented on dojo_plugin/pages/workspace.py:
```diff
@@ -46,14 +46,15 @@ def start_on_demand_service(user, service_name):
 def view_desktop():
     user_id = request.args.get("user")
     password = request.args.get("password")
+    url_only = request.args.get("url_only")
```
I'm working on it.

JensHouses replied:
https://github.com/pwncollege/dojo/pull/489/commits/5dffca24e842cc029537593dc1067c8b3ee27f0c should do it

ConnorNelson commented on dojo_plugin/pages/workspace.py:
```diff
@@ -20,88 +20,18 @@
     user = get_current_user()
     active = bool(get_current_dojo_challenge())
+    iframe_src = "" if service=="desktop" else f"/workspace/{service}/"
+    loadUrlAsync = service == "desktop"
```
Let's get rid of the decision to `loadUrlAsync`. All `/workspace/*` should be loaded async; the desktop is not special :)

JensHouses replied:
code just seams to have the iframe src /workspace/code/
what other workspaces are there ? :)

ConnorNelson commented on dojo_theme/templates/iframe.html:
```diff
@@ -45,6 +45,16 @@ <h1 class="inactive">No active challenge session; start a challenge!<h1>
 {{ super() }}
+{% if active and loadUrlAsync %}
+<script>
+    let response =  fetch("/pwncollege_api/v1/workspace").then(response => response.json()).then(data => {
```
Somehow we should pass to the workspace endpoint what url we're looking for. You can pass this all the way through with a jina `{{ service }}`.

Tangentially, it might also make sense to rename this from iframe.html => workspace.html. If there is someone else using iframe.html besides a workspace-related feature (I don't think so?) then we'd want to make a copy.

ConnorNelson commented on dojo_plugin/api/v1/workspace.py:
```diff
@@ -0,0 +1,66 @@
+@workspace_namespace.route("")
+class view_desktop(Resource):
+    @authed_only
+    def get(self):
```
Probably add a `service = request.args.get("service")` or something like that.

ConnorNelson reviewed (commented):
This looks super awesome! Just 1 more small change: let's support `async` for all workspace services, not just the desktop. Then it will be ready to merge. Thanks for all your work!

JensHouses commented:
Final changes: 🤣 
- Added `service` parameter to `view_desktop` endpoint in `workspace.py`.
- Updated iframe source handling based on `service` parameter in `workspace.py`.
- Simplified `view_workspace` route in `pages/workspace.py`.
- Removed async URL loading and share URLs script from `iframe.html`.
- Created new `workspace.html` template for dynamic iframe loading.

ConnorNelson commented:
Great! Looks like testcases aren't passing though?

JensHouses commented:
Spent 3 hours debugging an issue where Firefox refused to open WebSockets, which led me to believe there was a problem with my code. After exhausting all possible code-related issues, I finally tested it in Chrome, where it worked perfectly fine. Turns out, the problem was with Firefox all along, not my code. Once I realized this, the actual task—changing a '2' to a '3'—took just a few minutes. 3 hours of work to change a single number!

zardus commented:
DEFCON's over, so we'll get back to getting this merged in shortly :-)

Looks like we'll need to squash, cause rebasing will have conflicts.

JensHouses commented:
Finally... =)

zardus commented:
Thanks, @JensHouses!! @ConnorNelson thoughts on the API/page split? If you're happy, then we yolo?

JensHouses commented on dojo_theme/static/js/dojo/navbar.js:
```diff
@@ -1,3 +1,29 @@
+broadcast.onmessage = (event) => {
+    if (event.data.msg === 'New challenge started') {
+        if (window.location.pathname === '/workspace/code') {
+            window.location.reload();
```
i cant figure out why reloading the iframe does not work. as the iframe url seams to be the same.
reloading the page seam to solve the problem. But i currently don't know why.
Any advice would be welcome.

zardus commented:
![image](https://github.com/user-attachments/assets/d2fa6b9f-6cd2-4a63-a14c-812e059be971)

zardus commented:
thanks!!!!
